File: shared/src/domain.ts
// Domain / wire types shared between the backend (which produces them) and the frontend
// (which consumes them). These describe the shapes that actually cross the HTTP/WS boundary.
// Server-only fields (e.g. a participant's runner_token) are NOT here — they live in
// backend-local row types that extend these.

export type Kind = "human" | "agent";

// One of an agent's live statuses. Working = actively running a turn; Idle = connected,
// waiting; Sleeping = machine stopped to save cost; Waking = machine starting, not yet connected;
// Offline = a self-hosted agent whose device/daemon is not connected (the backend cannot wake it —
// queued work waits until the device comes back online). Offline is distinct from Sleeping, which
// the platform CAN wake on demand.
export type AgentStatus = "working" | "idle" | "sleeping" | "waking" | "offline";

// The provisioners an agent's runner can run under. 'docker'/'fly' are cloud sandboxes the backend
// owns; 'self_hosted' runs on a user's own registered device (see RunnerHost) via a daemon that
// dials the host-control channel (shared/src/host-protocol.ts).
export const RUNNER_PROVIDERS = ["docker", "fly", "self_hosted"] as const;
export type RunnerProvider = (typeof RUNNER_PROVIDERS)[number];

// The public shape of a participant row (everything persisted except server-only secrets like
// runner_token). This is what the backend serializes; the DB row type extends it server-side.
export interface ParticipantBase {
  id: string;
  kind: Kind;
  workspace_id: string; // the workspace this participant belongs to (Slack-style multi-tenancy)
  role: string; // membership role within the workspace: 'admin' | 'member'
  handle: string;
  display_name: string;
  repo: string | null;
  firebase_uid: string | null;
  email: string | null;
  avatar_url: string | null;
  model: string | null; // agent model override (null = agent-config default)
  mode: string; // an SDK permission mode: default|acceptEdits|plan|bypassPermissions|dontAsk
  effort: string; // reasoning effort: low|medium|high|xhigh (default 'medium'); see EFFORT_LEVELS
  runtime: string; // 'sdk' (all agents; legacy 'ma' rows may exist on old databases)
  // Context-window occupancy reported by the runner after each turn (null = no report yet).
  context_tokens: number | null;
  context_max_tokens: number | null;
  context_updated_at: string | null;
  runner_provider: string; // RunnerProvider — which Provisioner impl owns this agent's runner
  // Provider handles (Fly: {machineId, volumeId}; self_hosted: {hostId, host?: {hostname,…}}).
  runner_meta: Record<string, unknown> | null;
  // Creator-written role/personality injected into the agent's system prompt (agents; null = none).
  // The agent's MEMORY.md mirror is NOT here — it can be large, so clients fetch it on demand
  // via GET /api/agents/:id/memory.
  persona: string | null;
}

// A participant as sent to clients: the public row plus a live `status` (agents only, computed
// from the runner connection at serialization time — not persisted). `memory_changed_at` is
// client-side only: stamped when an agent_memory_changed broadcast lands, so an open profile's
// Memory section knows to refetch.
export interface Participant extends ParticipantBase {
  status?: AgentStatus;
  memory_changed_at?: string;
}

// --- Self-hosted devices (a registered machine that can run agents) ---

// Who, besides the owner, may assign an agent to run on a device. 'owner_only' (default) = only
// the account that registered it; 'workspace_members' = any member of a workspace the owner has
// shared the device into (shared_workspace_ids). Running an agent on a device is code execution
// with the owner's OS privileges, so this defaults closed.
export type DeviceAssignPolicy = "owner_only" | "workspace_members";

// A machine a user registered with `jungle-agents connect`. Account-scoped (owned by a Google
// account, selectable across that account's workspaces). `online` is derived at serialization
// time from whether the device's control connection is live; `running_agents` counts agents
// currently executing on it. Server-only fields (the device token hash) never appear here.
export interface RunnerHost {
  id: string;
  name: string; // user-editable; defaults to the hostname
  hostname: string | null;
  platform: string | null; // process.platform, e.g. 'darwin' | 'linux'
  arch: string | null; // process.arch, e.g. 'arm64' | 'x64'
  runner_version: string | null;
  assign_policy: DeviceAssignPolicy;
  shared_workspace_ids: string[]; // workspaces the device is shared into (workspace_members policy)
  // Whether agents on this device run in an isolated per-agent workspace (true, the default) or
  // directly in the directory `jungle-agents connect` was run from (false). False = the agent has
  // the user's real files in its cwd; per-agent state (memory/session/git-creds) stays isolated.
  sandboxed: boolean;
  created_at: string;
  last_seen_at: string | null;
  online: boolean; // derived: control channel currently connected
  running_agents: number; // derived: agents with a live runner on this device
}

// --- Workspaces (Slack-style multi-tenancy) ---

// A workspace as sent to clients.
export interface Workspace {
  id: string;
  name: string;
}

// The Google identity behind a signed-in user (before/independent of any workspace participant).
export interface GoogleProfile {
  uid: string;
  email: string | null;
  name: string | null;
  picture: string | null;
}

// One of a signed-in account's workspace memberships: the workspace, the account's participant in
// it (handle/role/etc.), and whether GitHub is connected for that participant.
export interface Membership {
  workspace: Workspace;
  participant: Participant;
  github: { connected: boolean; login?: string };
}

// GET /api/me: the signed-in Google account and every workspace it belongs to. `suggestedHandle`
// seeds the handle field when creating/joining a workspace.
export interface Me {
  profile: GoogleProfile;
  memberships: Membership[];
  suggestedHandle: string;
}

// GET /api/invites/:token: what a would-be joiner sees before accepting an invite link.
export interface InviteInfo {
  valid: boolean; // false = unknown / revoked / expired token
  workspaceName?: string;
  alreadyMember?: boolean; // the signed-in account is already in this workspace
}

// The attachment fields stored on an upload. A signed download `url` is added at the edge by
// the backend (attachments.withUrls) to produce `Attachment` — the url is never stored.
export interface AttachmentMeta {
  id: string;
  filename: string;
  mime: string;
  size_bytes: number;
  width: number | null;
  height: number | null;
}

// An attachment as sent to clients: stored metadata plus an origin-relative signed download
// path (prefix with the API origin to fetch).
export interface Attachment extends AttachmentMeta {
  url: string;
}

export interface Message {
  id: string;
  channel_id: string;
  seq: string; // bigint serialized as string
  sender_id: string;
  sender_handle: string;
  body: string;
  created_at: string;
  cascade_budget: number | null;
  // Agent messages: the runner turn that produced this message (null for human messages and
  // for agent sends outside a tracked turn). Drives the "view the work" affordance.
  turn_id: string | null;
  // Threads: null on top-level messages; the root message's id on replies. also_to_channel
  // marks a reply that was also echoed into the main timeline. reply_count/last_reply_at are
  // denormed on the ROOT (0/null elsewhere) and drive the "N replies" footer + Threads view.
  thread_root_id: string | null;
  also_to_channel: boolean;
  reply_count: number;
  last_reply_at: string | null;
  mentions: { id: string; handle: string }[];
  attachments: AttachmentMeta[];
}

// A message as sent to clients (WS `message` frame + REST message responses): like Message,
// but each attachment carries a signed download `url` (added at the edge by withUrls).
export interface WireMessage extends Omit<Message, "attachments"> {
  attachments: Attachment[];
}

// A channel/DM row as returned by the channel list (GET /api/channels).
export interface ChannelListItem {
  id: string;
  name: string;
  kind: string;
  dm_with: string | null; // for dm channels: the other member's handle
  unread_count: number; // messages after the requester's last_read_seq, excluding their own
  has_mention: boolean; // any unread message @mentions the requester
  member_agent_ids: string[]; // agent members of this channel
}

// A followed thread with unread replies, for the "Threads" view (GET /api/threads/unread).
export interface UnreadThread {
  root_id: string;
  channel_id: string;
  channel_name: string;
  root_sender_handle: string;
  root_body: string;
  reply_count: number;
  last_reply_at: string | null;
  unread_count: number; // replies after the requester's thread last_read_seq, excluding their own
}

// A single Claude Agent SDK stream message persisted for an sdk agent. `event` is the raw SDK
// message JSON (system/assistant/user/result); render defensively (shapes may vary).
export interface AgentEvent {
  id: number;
  turn_id: string;
  event: unknown;
  created_at: string;
}

// The link kinds recognized as durable work artifacts (see backend services/deliverables.ts —
// classification lives there; this union is the wire vocabulary).
export type DeliverableKind =
  | "github_pr"
  | "github_issue"
  | "github"
  | "notion"
  | "google_doc"
  | "google_drive"
  | "linear"
  | "granola";

// A durable work artifact an agent produced (a PR opened, a doc written, …), extracted from the
// links in its messages. Powers the Deliverables feed + inline artifact cards.
export interface Deliverable {
  id: number;
  agent_id: string;
  agent_handle: string;
  channel_id: string;
  channel_name: string;
  channel_kind: string;
  message_id: string;
  kind: DeliverableKind;
  title: string | null;
  url: string;
  created_at: string;
}

// One message search hit (GET /api/search), scoped to channels the requester belongs to.
export interface SearchResult {
  message_id: string;
  channel_id: string;
  channel_name: string;
  channel_kind: string;
  dm_with: string | null; // for dm channels: the other member's handle
  thread_root_id: string | null;
  sender_handle: string;
  body: string;
  created_at: string;
}

File: ios/Jungle/Wire/Domain.swift
import Foundation

// Hand-ported from shared/src/domain.ts — the wire shapes that cross the HTTP/WS boundary.
// Keep 1:1 with the TypeScript source; a PR touching domain.ts should touch this file.
// Domain rows are snake_case on the wire (explicit CodingKeys); timestamps stay ISO strings
// (parse at display time, like the web).

enum Kind: String, Codable, Sendable {
    case human
    case agent
}

// Decoded tolerantly: an unrecognized status from a newer server maps to .offline-ish unknown
// rather than failing the whole participant decode.
enum AgentStatus: String, Codable, Sendable {
    case working
    case idle
    case sleeping
    case waking
    case offline
    case unknown

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = AgentStatus(rawValue: raw) ?? .unknown
    }
}

enum RunnerProvider: String, Codable, Sendable {
    case docker
    case fly
    case selfHosted = "self_hosted"
}

// ParticipantBase + Participant from domain.ts, flattened (Swift has no interface extension;
// `status`/`memoryChangedAt` are simply optional here).
struct Participant: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var kind: Kind
    var workspaceId: String
    var role: String
    var handle: String
    var displayName: String
    var repo: String?
    var firebaseUid: String?
    var email: String?
    var avatarUrl: String?
    var model: String?
    var mode: String
    var effort: String
    var runtime: String
    var contextTokens: Int?
    var contextMaxTokens: Int?
    var contextUpdatedAt: String?
    var runnerProvider: String
    var runnerMeta: [String: JSONValue]?
    var persona: String?
    var status: AgentStatus?
    // Client-side only on the web (stamped when agent_memory_changed lands); optional here too.
    var memoryChangedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, kind, role, handle, repo, email, model, mode, effort, runtime, persona, status
        case workspaceId = "workspace_id"
        case displayName = "display_name"
        case firebaseUid = "firebase_uid"
        case avatarUrl = "avatar_url"
        case contextTokens = "context_tokens"
        case contextMaxTokens = "context_max_tokens"
        case contextUpdatedAt = "context_updated_at"
        case runnerProvider = "runner_provider"
        case runnerMeta = "runner_meta"
        case memoryChangedAt = "memory_changed_at"
    }
}

// --- Self-hosted devices ---

enum DeviceAssignPolicy: String, Codable, Sendable {
    case ownerOnly = "owner_only"
    case workspaceMembers = "workspace_members"
}

struct RunnerHost: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
    var hostname: String?
    var platform: String?
    var arch: String?
    var runnerVersion: String?
    var assignPolicy: DeviceAssignPolicy
    var sharedWorkspaceIds: [String]
    var sandboxed: Bool
    var createdAt: String
    var lastSeenAt: String?
    var online: Bool
    var runningAgents: Int

    enum CodingKeys: String, CodingKey {
        case id, name, hostname, platform, arch, sandboxed, online
        case runnerVersion = "runner_version"
        case assignPolicy = "assign_policy"
        case sharedWorkspaceIds = "shared_workspace_ids"
        case createdAt = "created_at"
        case lastSeenAt = "last_seen_at"
        case runningAgents = "running_agents"
    }
}

// The minimum runner version that honors the `sandboxed` flag (see domain.ts).
let unsandboxedMinRunnerVersion = "0.1.1"

// Compare two dotted numeric version strings; missing/non-numeric parts count as 0.
func compareVersions(_ a: String, _ b: String) -> Int {
    let pa = a.split(separator: ".").map { Int($0) ?? 0 }
    let pb = b.split(separator: ".").map { Int($0) ?? 0 }
    for i in 0..<max(pa.count, pb.count) {
        let d = (i < pa.count ? pa[i] : 0) - (i < pb.count ? pb[i] : 0)
        if d != 0 { return d }
    }
    return 0
}

func supportsUnsandboxed(_ version: String?) -> Bool {
    guard let version, !version.isEmpty else { return false }
    return compareVersions(version, unsandboxedMinRunnerVersion) >= 0
}

// --- Workspaces ---

struct Workspace: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
}

struct GoogleProfile: Codable, Hashable, Sendable {
    var uid: String
    var email: String?
    var name: String?
    var picture: String?
}

struct Membership: Codable, Hashable, Sendable {
    var workspace: Workspace
    var participant: Participant
    var github: GithubStatus

    struct GithubStatus: Codable, Hashable, Sendable {
        var connected: Bool
        var login: String?
    }
}

// GET /api/me
struct Me: Codable, Hashable, Sendable {
    var profile: GoogleProfile
    var memberships: [Membership]
    var suggestedHandle: String
}

// GET /api/invites/:token
struct InviteInfo: Codable, Hashable, Sendable {
    var valid: Bool
    var workspaceName: String?
    var alreadyMember: Bool?
}

// An admin-visible invite link row.
struct Invite: Codable, Hashable, Sendable {
    var token: String
    var expiresAt: String?
    var createdAt: String

    enum CodingKeys: String, CodingKey {
        case token
        case expiresAt = "expires_at"
        case createdAt = "created_at"
    }
}

// --- Attachments / messages ---

// AttachmentMeta + Attachment flattened: `url` (origin-relative signed download path) is present
// on client-facing shapes (WireMessage attachments, upload response) and absent otherwise.
struct Attachment: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var filename: String
    var mime: String
    var sizeBytes: Int
    var width: Int?
    var height: Int?
    var url: String?

    enum CodingKeys: String, CodingKey {
        case id, filename, mime, width, height, url
        case sizeBytes = "size_bytes"
    }
}

// WireMessage from domain.ts (a message as sent to clients; attachments carry signed urls).
struct Message: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var channelId: String
    var seq: String // bigint serialized as string
    var senderId: String
    var senderHandle: String
    var body: String
    var createdAt: String
    var cascadeBudget: Int?
    var turnId: String?
    var threadRootId: String?
    var alsoToChannel: Bool
    var replyCount: Int
    var lastReplyAt: String?
    var mentions: [Mention]
    var attachments: [Attachment]

    struct Mention: Codable, Hashable, Sendable {
        var id: String
        var handle: String
    }

    enum CodingKeys: String, CodingKey {
        case id, seq, body, mentions, attachments
        case channelId = "channel_id"
        case senderId = "sender_id"
        case senderHandle = "sender_handle"
        case createdAt = "created_at"
        case cascadeBudget = "cascade_budget"
        case turnId = "turn_id"
        case threadRootId = "thread_root_id"
        case alsoToChannel = "also_to_channel"
        case replyCount = "reply_count"
        case lastReplyAt = "last_reply_at"
    }
}

// ChannelListItem from domain.ts. Several fields are optional because POST /api/channels
// returns a bare channel row (see frontend Channel type, which marks them optional).
struct Channel: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
    var kind: String
    var dmWith: String?
    var unreadCount: Int?
    var hasMention: Bool?
    var memberAgentIds: [String]?

    enum CodingKeys: String, CodingKey {
        case id, name, kind
        case dmWith = "dm_with"
        case unreadCount = "unread_count"
        case hasMention = "has_mention"
        case memberAgentIds = "member_agent_ids"
    }

    var isDM: Bool { kind == "dm" }
}

// GET /api/threads/unread
struct UnreadThread: Codable, Hashable, Sendable {
    var rootId: String
    var channelId: String
    var channelName: String
    var rootSenderHandle: String
    var rootBody: String
    var replyCount: Int
    var lastReplyAt: String?
    var unreadCount: Int

    enum CodingKeys: String, CodingKey {
        case rootId = "root_id"
        case channelId = "channel_id"
        case channelName = "channel_name"
        case rootSenderHandle = "root_sender_handle"
        case rootBody = "root_body"
        case replyCount = "reply_count"
        case lastReplyAt = "last_reply_at"
        case unreadCount = "unread_count"
    }
}

// One persisted Claude Agent SDK stream message. `event` is raw SDK JSON — render defensively.
// `id` is a bigint that the REST API serializes as a STRING while live frames use synthetic
// numbers (Date.now()+random on web) — decode both.
struct AgentEvent: Codable, Identifiable, Hashable, Sendable {
    var id: Double
    var turnId: String
    var event: JSONValue
    var createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, event
        case turnId = "turn_id"
        case createdAt = "created_at"
    }

    init(id: Double, turnId: String, event: JSONValue, createdAt: String) {
        self.id = id
        self.turnId = turnId
        self.event = event
        self.createdAt = createdAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if let n = try? c.decode(Double.self, forKey: .id) {
            id = n
        } else {
            let s = try c.decode(String.self, forKey: .id)
            id = Double(s) ?? 0
        }
        turnId = try c.decodeIfPresent(String.self, forKey: .turnId) ?? "—"
        event = try c.decodeIfPresent(JSONValue.self, forKey: .event) ?? .null
        createdAt = try c.decode(String.self, forKey: .createdAt)
    }
}

// --- Deliverables ---

enum DeliverableKind: String, Codable, Sendable {
    case githubPr = "github_pr"
    case githubIssue = "github_issue"
    case github
    case notion
    case googleDoc = "google_doc"
    case googleDrive = "google_drive"
    case linear
    case granola
    case unknown

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = DeliverableKind(rawValue: raw) ?? .unknown
    }
}

struct Deliverable: Codable, Identifiable, Hashable, Sendable {
    var id: Int
    var agentId: String
    var agentHandle: String
    var channelId: String
    var channelName: String
    var channelKind: String
    var messageId: String
    var kind: DeliverableKind
    var title: String?
    var url: String
    var createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, kind, title, url
        case agentId = "agent_id"
        case agentHandle = "agent_handle"
        case channelId = "channel_id"
        case channelName = "channel_name"
        case channelKind = "channel_kind"
        case messageId = "message_id"
        case createdAt = "created_at"
    }
}

// GET /api/search
struct SearchResult: Codable, Hashable, Sendable {
    var messageId: String
    var channelId: String
    var channelName: String
    var channelKind: String
    var dmWith: String?
    var threadRootId: String?
    var senderHandle: String
    var body: String
    var createdAt: String

    enum CodingKeys: String, CodingKey {
        case body
        case messageId = "message_id"
        case channelId = "channel_id"
        case channelName = "channel_name"
        case channelKind = "channel_kind"
        case dmWith = "dm_with"
        case threadRootId = "thread_root_id"
        case senderHandle = "sender_handle"
        case createdAt = "created_at"
    }
}
